Sources/MongoSwift/ChangeStream.swift

```diff
@@ -90,60 +105,101 @@ public class ChangeStream<T: Codable>: Sequence, IteratorProtocol {
+        }
+        self.resumeToken = ResumeToken(resumeToken)
+        return try self.decoder.decode(T.self, from: doc)
+    }
+
+    /// Destroys the underlying mongoc change stream and closes the state in a blocking manner.
+    private func blockingClose() {
+        guard case let .open(changeStream, _, _) = self.state else {
+            return
+        }
+        self.pollingSemaphore.wait()
+        mongoc_change_stream_destroy(changeStream)
+        self.state = .closed
+        self.pollingSemaphore.signal()
+    }
+
     /// Returns the next `T` in the change stream or nil if there is no next value. Will block for a maximum of
     /// `maxAwaitTimeMS` milliseconds as specified in the `ChangeStreamOptions`, or for the server default timeout
     /// if omitted.
-    public func next() -> T? {
-        // We already closed the mongoc change stream, either because we reached the end or encountered an error.
-        guard case let .open(_, connection, client, session) = self.state else {
-            self.error = ClosedChangeStreamError
-            return nil
+    public func next() -> EventLoopFuture<T?> {
+        let operation = NextOperation(target: .changeStream(self))
+        return self.client.executeOperationAsync(operation)
+    }
+
+    /// Returns all of the events seen by this change stream so far.
+    public func all() -> EventLoopFuture<[T]> {
+        let operation = AllOperation(target: .changeStream(self))
+        return self.client.executeOperationAsync(operation)
+    }
+
+    /// Executes the provided closure against each event seen by the change stream.
+    /// This will cause the stream to continuously poll for new events in the background. Once this method has been
+    /// called, other methods that check for events (e.g. all, next, forEach) should not be used.
+    public func forEach(body: @escaping (Result<T, Error>) -> Void) {
+        self.pollingSemaphore.wait()
+
+        guard self.isAlive else {
+            self.pollingSemaphore.signal()
+            return
         }
-        do {
-            let operation = NextOperation(target: .changeStream(self))
-            guard let out = try client.executeOperation(operation, using: connection, session: session) else {
-                self.error = self.getChangeStreamError()
-                if self.error != nil {
-                    self.close()
+
+        self.next().whenComplete { result in
+            switch result {
+            case let .success(event):
+                if let event = event {
+                    body(.success(event))
                 }
-                return nil
+            case let .failure(error):
+                body(.failure(error))
+                return
             }
-            return out
-        } catch {
-            self.error = error
-            self.close()
-            return nil
+            self.pollingSemaphore.signal()
+            self.forEach(body: body)
```

Comment:
This method will likely be very long-lived, which means these recursion calls can add up in terms of memory cost. I'm not really sure how to get around this, since any long term polling without async/await will need recursion, though.